### PR TITLE
Upgrade to SpringBoot 2.6.1, updating unit test and code accordingly

### PR DESCRIPTION
**Pull request by mr-c**
(redo of https://github.com/common-workflow-language/cwlviewer/pull/280 )

The reported issue is related to the following case: 

https://stackoverflow.com/questions/49316751/spring-data-jpa-findone-change-to-optional-how-to-use-this

Solution:

[Editing Pom.xml]

update Spring Boot:

<parent>
   <groupId>org.springframework.boot</groupId>
   <artifactId>spring-boot-starter-parent</artifactId>
   <version>2.2.2.RELEASE</version>
   <relativePath/>
</parent>

add the following dependency for ensuring that we use the appropriate spring-data-commons version. 

<dependency>
   <groupId>org.springframework.data</groupId>
   <artifactId>spring-data-commons</artifactId>
   <version>2.0.1.RELEASE</version>
</dependency>

[WorkflowService.java]

Replace any occurrence of the findOne with findById(id).orElse(null) here:

org.commonwl.view.workflow.WorkflowRepository
org.commonwl.view.workflow.QueuedWorkflowRepository

With these changes we should expect to build successfully. 

- [x] upgrade to Spring Boot `2.6.1` (latest)
- [x] upgrade to JUnit 5 (comes with Spring Boot, unless we enable the JUnit 4 integration layer)
- [x] Add a javax.validation implementation
- [x] Upgrade `spring-data-commons` to `2.6.0`
- [x] Fix unit tests
- [x] Upgrade Jena
- [x] Manual testing to confirm everything works
- [x] Fix GH Actions build

### CODE REVIEW

**Comment by mr-c** [comment]
> Error creating bean with name 'mongoConfig': Unsatisfied dependency expressed through field 'mongoMappingContext'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'mongoMappingContext' defined in class path resource [org/springframework/boot/autoconfigure/data/mongo/MongoDataConfiguration.class]: Initialization of bean failed; nested exception is java.lang.NoSuchMethodError: org.springframework.data.mapping.context.AbstractMappingContext.setApplicationContext(Lorg/springframework/context/ApplicationContext;)V

https://travis-ci.com/github/common-workflow-language/cwlviewer/builds/198507618#L1273

**Comment by mr-c** [comment]
@kinow I assigned you this PR, but feel free to make a new branch

**Comment by kinow** [comment]
Rebased.

File: src/test/java/org/commonwl/view/PageControllerTest.java [hunk 1]
```diff
@@ -19,8 +19,8 @@
 
 package org.commonwl.view;
 
-import org.junit.Before;
-import org.junit.Test;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
```
Comment (kinow) [inline src/test/java/org/commonwl/view/PageControllerTest.java:23]: This is part of JUnit4 -> JUnit 5, which is a transitive dependency of SpringBoot.

File: pom.xml [hunk 2]
```diff
@@ -124,6 +124,11 @@
 			<artifactId>spring-data-commons</artifactId>
 			<version>2.0.1.RELEASE</version>
 		</dependency>
+		<dependency>
+			<groupId>org.hibernate.validator</groupId>
+			<artifactId>hibernate-validator</artifactId>
+			<version>6.0.13.Final</version>
+		</dependency>
```
Comment (kinow) [inline pom.xml:131]: Looks like `javax.validation` implementation is not included, so had to import the Hibernate validator (this is not ORM, just the validation part, which is a reference implementation of JSR-303)

**Comment by kinow** [comment]
- [x] upgrade to Spring Boot 2.6.1 (latest)
- [x] upgrade to JUnit 5 (comes with Spring Boot, unless we enable the JUnit 4 integration layer)
- [x] Add a javax.validation implementation
- [ ] Fix unit tests (90%)
- [ ] Manual testing to confirm everything works
- [ ] Upgrade Commons Compress
- [ ] Upgrade Jena

**Comment by mr-c** [comment]
Thanks @kinow ! I moved your checklist to the first comment so that it shows up on the issues browser. Let me know if you don't have the permissions to edit it further.

**Comment by kinow** [comment]
> Thanks @kinow ! I moved your checklist to the first comment so that it shows up on the issues browser. Let me know if you don't have the permissions to edit it further.

That'll be easier to update and for others to see too. Thanks @mr-c !

**Comment by kinow** [comment]
Huh, funny, `mvnw test` failed on GH Actions, but passing locally. Will upgrade the remaining libraries, have a look at `miniwdl` + types, and revisit this PR later (to give my :brain: a break :smile: )

**Comment by kinow** [comment]
Manual tests passed with `docker-compose up --build`. Unit tests passing locally. Last part now, fix test in GH Actions, then it should be ready for review! :tada:

**Comment by kinow** [comment]
All right. The MongoDB test passed locally, because the tests were run sequentially. On GitHub Actions they were running in parallel, so there was a conflict with ports. Setting the port number to `0` makes SpringData/MongoDB to allocate a random port :+1:

**Comment by kinow** [comment]
@mr-c this PR should upgrade SpringBoot to `2.6.1`, the latest release. The API changed a little from 1.5.x to 2.x, and also to 2.6.1. Removed some deprecated code, and fixed settings and Java annotations so that all tests would pass. Tested with the `docker-compose.yml` file, after replacing `image` by the commented `build`.

I don't know enough about the production environment to say it will work flawlessly when merged & deployed. So, if possible, a backup when upgrading would be recommended IMO.

Let me know if you know what would be the next steps here, or if I need to talk with someone else to see it merged & deployed :tada: 

-Bruno

**Comment by mr-c** [comment]
@kinow this is fantastic! Deployment is manual, so no worries about merging this PR. Do you want me to squash this down to one commit, or do you want to cleanup these commits and then I'll rebase onto the primary branch?

File: pom.xml [hunk 3]
```diff
@@ -88,36 +93,57 @@
...
 		<dependency>
 			<groupId>com.github.jsonld-java</groupId>
 			<artifactId>jsonld-java</artifactId>
 			<version>${ver.jsonldjava}</version>
 		</dependency>
+		<!-- Git in Java -->
 		<dependency>
 			<groupId>org.eclipse.jgit</groupId>
 			<artifactId>org.eclipse.jgit</artifactId>
 			<version>5.13.0.202109080827-r</version>
```
Comment (kinow) [inline pom.xml:117]: Bumping this to 6.x breaks the build. I think jgit 6.x requires Java 11, but our Docker+Docker Compose image didn't work when I tested upgrading it. For a follow-up/later.

**Comment by kinow** [comment]
> @kinow this is fantastic! Deployment is manual, so no worries about merging this PR. Do you want me to squash this down to one commit, or do you want to cleanup these commits and then I'll rebase onto the primary branch?

Forgot about squashing commits. Done now :slightly_smiling_face: Thanks!

**Comment by mr-c** [comment]
Oh, I can always easily squash down to one commit myself in the web interface. I wondered if you preferred one big commit or a series of fewer commits with cleaned up descriptions.

File: pom.xml (hunk 3 above)
Comment (mr-c) [reply pom.xml:117]: Can you make an issue for this?

Comment (kinow) [reply pom.xml:117]: Good idea, done! https://github.com/common-workflow-language/cwlviewer/issues/367

**Comment by mr-c** [comment]
@kinow I force pushed with correct attribution. You wrote a lot of nice commit messages, and some of them should be saved. But since I didn't have a local checkout prior to your squash I can't regroup your original commits. Would you mind doing so? (next week is fine!)

I did make a local backup of view.commonwl.org via the instructions at https://github.com/common-workflow-language/cwlviewer#dumprestore and I did a partial restore until I got bored. Seems fine!

**Comment by kinow** [comment]
> @kinow I force pushed with correct attribution.

Did I replace an author somewhere, or forgot a commit? I thought perhaps you meant the co-author bit that appears after I rebase; but I don't know how to remove it. Hopefully I didn't mess it up now...

>You wrote a lot of nice commit messages, and some of them should be saved. But since I didn't have a local checkout prior to your squash I can't regroup your original commits. Would you mind doing so? (next week is fine!)

Thanks! I remember adding some links in the commits to explain where settings/changes were coming from. But there are two commits fixing unit tests which could be combined perhaps. The commit message could be merged into a single message with links & notes.

Feel free to re-organize the commits if you'd like @mr-c. I used `git reflog` to reset to before my rebase, and cherry-pick your commits. I hope I didn't miss anything.

> I did make a local backup of view.commonwl.org via the instructions at https://github.com/common-workflow-language/cwlviewer#dumprestore and I did a partial restore until I got bored. Seems fine!

Hooray! :tada: 

Thanks!

**Comment by mr-c** [comment]
> > @kinow I force pushed with correct attribution.
> 
> Did I replace an author somewhere, or forgot a commit? I thought perhaps you meant the co-author bit that appears after I rebase; but I don't know how to remove it.



No, the issue was your commit *didn't* list you as an author! :-)

> Hopefully I didn't mess it up now...
Everything looks good, thanks!

> Feel free to re-organize the commits if you'd like @mr-c. I used `git reflog` to reset to before my rebase, and cherry-pick your commits. I hope I didn't miss anything.

I did a local check and all is good; thanks! I'll do some light commit combining with `git rebase -i` and I'll force push

**Comment by kinow** [comment]
>No, the issue was your commit didn't list you as an author! :-)

:rofl: ops!

Thanks @mr-c !

**Pull request merged**